services/index/internal/model/profile_test.go

```diff
@@ -0,0 +1,266 @@
+	tests := []struct {
+		name             string
+		profileStr       string
+		tagsArraySize    int
+		tagsStringLength int
+		expected         map[string]interface{}
+		expectErr        bool
+	}{
+		{
+			name:             "valid tags filtering",
+			profileStr:       `{"tags": ["tag1", "tag2", "tag3"]}`,
+			tagsArraySize:    3,
+			tagsStringLength: 10,
+			expected: map[string]interface{}{
+				"tags": []string{"tag1", "tag2", "tag3"},
+			},
+			expectErr: false,
+		},
+		{
+			name:             "invalid tags filtering, too many tags",
+			profileStr:       `{"tags": ["tag1", "tag2", "tag3"]}`,
+			tagsArraySize:    2,
+			tagsStringLength: 10,
+			expected: map[string]interface{}{
+				"tags": []string{"tag1", "tag2"},
+			},
+			expectErr: false,
+		},
+		{
+			name:             "valid tags filtering, truncate long tag",
+			profileStr:       `{"tags": ["verylongtagwithmorethan10chars", "tag2"]}`,
+			tagsArraySize:    2,
+			tagsStringLength: 10,
+			expected: map[string]interface{}{
+				"tags": []string{"verylongta", "tag2"},
+			},
+			expectErr: false,
+		},
+		{
+			name:             "invalid profileStr",
+			profileStr:       `{tags: ["tag1", "tag2"]}`, // incorrect JSON syntax
+			tagsArraySize:    3,
+			tagsStringLength: 10,
+			expected:         map[string]interface{}{},
+			expectErr:        true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			config.Values.Server.TagsArraySize = strconv.Itoa(tt.tagsArraySize)
+			config.Values.Server.TagsStringLength = strconv.Itoa(
+				tt.tagsStringLength,
+			)
+
+			profile := model.NewTestProfile(tt.profileStr)
+			err := profile.FilterTags()
+			if tt.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, tt.expected, profile.GetJSON())
+			}
+		})
+	}
+}
+
+func TestValidatePrimaryURL(t *testing.T) {
+	tests := []struct {
+		name       string
+		primaryURL string
+		expected   map[string]interface{}
+		expectErr  bool
+	}{
+		{
+			name:       "valid URL",
+			primaryURL: "https://example.com/path?query=param",
+			expected: map[string]interface{}{
+				"primary_url": "example.com/path?query=param",
+			},
+			expectErr: false,
+		},
+		{
+			name:       "valid URL with www",
+			primaryURL: "https://www.example.com/path?query=param",
+			expected: map[string]interface{}{
+				"primary_url": "example.com/path?query=param",
+			},
+			expectErr: false,
+		},
+		{
+			name:       "valid URL without scheme",
+			primaryURL: "example.com/path?query=param",
+			expected: map[string]interface{}{
+				"primary_url": "example.com/path?query=param",
+			},
+			expectErr: false,
+		},
+		{
+			name:       "invalid URL",
+			primaryURL: "://example.com",
+			expected: map[string]interface{}{
+				"primary_url": "",
+			},
+			expectErr: false,
+		},
+		{
+			name:       "no primary_url",
+			primaryURL: "",
+			expected: map[string]interface{}{
+				"primary_url": "",
+			},
+			expectErr: false,
```

Comment:
@geoffturk Should we filter out invalid values, or should they return an error?

Reply:
@maxsarratt Good catch. It should return an error. This is a bug.

<img width="591" alt="image" src="https://github.com/MurmurationsNetwork/MurmurationsServices/assets/32190662/f24e635e-04d6-4467-b04d-c9bd8288293c">

Reply:
I have fixed the issue. 👍

Reply:
What should we do if we encounter an error while trying to set the node as valid? Should we send it to another state? It seems like our original implementation has some issues, which is why it gets stuck in the loop.

Reply:
I think we need to acknowledge the message even if we encounter an error. Otherwise, NATS will keep sending the same message.

Reply:
There's an issue with AddSync api. If we fail to add a node, it will keep looping until it gets the result...